### PR TITLE
Wall tests

### PR DESCRIPTION
**Pull request by rykerfish**
Adding some tests for DPStokes (and maybe other solvers later) that compare to some [reference results](https://github.com/stochasticHydroTools/DoublyPeriodicStokes/tree/main/figures).

### CODE REVIEW

File: solvers/DPStokes/mobility.h
Comment (rykerfish) [inline solvers/DPStokes/mobility.h:1]: @RaulPPelaez does this change look correct? In my new test, the self mobility didn't go to zero at the bottom wall without this change. I suspect it's supposed to add a buffer to the top boundary instead of the bottom wall since a comment in the area says to add a buffer for an open boundary.

File: solvers/DPStokes/mobility.h [hunk 1]
```diff
@@ -64,7 +64,7 @@ class DPStokes: public libmobility::Mobility{
       this->dppar.zmin -= this->dppar.w*h/2;
     }
     if(this->wallmode == "bottom"){
-      this->dppar.zmin -= this->dppar.w*h/2;
+      this->dppar.zmax += this->dppar.w*h/2;
     }
```
Comment (rykerfish) [inline solvers/DPStokes/mobility.h:68]: @RaulPPelaez does this change look correct? In my new test, the self mobility didn't go to zero at the bottom wall without this change. I suspect it's supposed to add a buffer to the top boundary instead of the bottom wall since a comment in the area says to add a buffer for an open boundary.

Comment (RaulPPelaez) [reply solvers/DPStokes/mobility.h:68]: yes indeed

Comment (RaulPPelaez) [reply solvers/DPStokes/mobility.h:68]: yes indeed, good catch.

**Comment by rykerfish** [comment]
@RaulPPelaez This plot is to illustrate an issue I've been running into with DPStokes in libMobility. Here, the square points and the lines are reference data from the [DPStokes repository](https://github.com/stochasticHydroTools/DoublyPeriodicStokes), and the triangular points are from libMobility generated from the bottom wall test on this branch. In general, libMobility lines up well, but the inset shows a small error between libMobility and the reference that is consistently of order 1e-3. The error goes to zero as $z/R_h \to 0$. This same mismatch occurs in slit channel examples as well. 

Brennan suspects this mismatch may be from suboptimal grid/kernel parameters. Do you agree? I'm going to attempt to integrate the [parameter selection](https://github.com/stochasticHydroTools/DoublyPeriodicStokes/blob/main/source/cpu/python/GridAndKernelConfig.py) from DPStokes to see if that resolves the mismatch. 

Let me know if have a different idea of what could be causing this systematic underestimation so I can check that out as well!

image10

**Comment by RaulPPelaez** [comment]
I agree with Brennan, it is most probably suboptimal grid parameters.

**Comment by RaulPPelaez** [comment]
We should be merging this before #19

**Comment by rykerfish** [comment]
@RaulPPelaez I haven't merged this yet as there are still problems with the DPStokes results not agreeing perfectly with reference results and Brennan and I think it should be possible to get the results to match to more than a tolerance of ~1e-3. Even when using the parameter selection code, there remains a persistent underestimation on the order of 1e-4.  

I have the reference code running from the DPStokes repository. As far as I can tell, all the grid and kernel parameters are the same between the two simulations. Do you know if there is anything that got changed between the libMobility version of DPStokes and the original version that could be causing this mismatch? We aren't sure where to look to try and resolve this as I've (mostly) ruled out the difference in parameters.

**Comment by rykerfish** [comment]
Here's a relative error plot for the xx and zz trans-trans components of the mobility matrix, computed as $\displaystyle \big\vert \frac{ref-libm}{ref} \big\vert$, versus z/R_h with R_h=1. The dashed vertical line is z=1. The error is 0 at z=0 for both bottom and and slit channel geometries but small but non-zero elsewhere. There's a strange local minimum at z=1.27 for the slit channel geometry where the error goes down to $O(1e^{-6})$ that I can't explain. 

The plot is the same for double and single precision, with negligible changes in the exact values of the computed mobilities. 

Here, the values for $\beta$, $w$, $\alpha$, $N_x, N_y, N_z$, and $h$ are the same as given by the UAMMD printout for libMobility and the parameter selection printout for the original DPStokes code.

param_select_residuals_single_precision

**Comment by RaulPPelaez** [comment]
Between the version from the reference and the current one used in libmobility there is this big commit in UAMMD https://github.com/RaulPPelaez/UAMMD/commit/c662443254bffefc4d3f23e74c9475fa79c07d80
You could try to undo this commit and try, maybe I messed up something subtle.
The version in the reference (which I understood you have ran and confirmed the results are consistent with the paper) is this https://github.com/RaulPPelaez/UAMMD/commits/7f6119fa03986d3ef03c46533f75e5a69a81340a/ . You could also try taking UAMMD to the latest version in the reference repo and see if it still reproduces the results. Just to confirm/discard this difference is rooted in some change to UAMMD since the paper.

The reference uses w=6 or 12, but this PR has w=4, I reckon that is taken into account?

**Comment by rykerfish** [comment]
@RaulPPelaez finally think I figured out the rest of the issues. Turns out that all the reference data from the DPStokes repository was generated using the CPU implementations. Additionally, although I was using the reference data for the w=4 kernel, their script for the pair mobility had a bug that could force the wrong kernel to be used. I regenerated all the data in that repository using the GPU implementations and with correct kernels and ended up with the relative error plots below, which Brennan and I are satisfied with. 

The pair mobility plot (2nd plot) isn't great since I didn't mess with the formatting enough, so a few notes on that one: The green curve with the error spikes is for two particles spaced 8R_h apart. The errors in that curve are consistently higher than the rest, but mostly small for z/R_h > 1. The large spike in the bottom right plot is because that height is at the exact middle of the slit channel, so the mobility gets small quickly. The actual error there is on the scale of 1e-9, but the relative error appears larger. 

To get these small relative errors, I had to use the exact same parameters as the DPStokes repository code. To avoid doing that in the future, I regenerated all the reference data using our implementations with our current default parameters in single precision to use as the reference going forwards. 

If all this looks good, I think this branch can finally get merged. Let me know what you think!


selfMobilityRelErr

pairMobilityRelErr

File: solvers/DPStokes/mobility.h [hunk 2]
```diff
@@ -51,9 +51,9 @@ class DPStokes: public libmobility::Mobility{
...
     this->dppar.hydrodynamicRadius = ipar.hydrodynamicRadius[0];
-    this->dppar.w = 6;
-    this->dppar.beta = 1.714*this->dppar.w;
-    real h = this->dppar.hydrodynamicRadius/1.554;
+    this->dppar.w = 4;
+    this->dppar.beta = 1.785*this->dppar.w;
+    real h = this->dppar.hydrodynamicRadius/1.205;
     this->dppar.alpha = this->dppar.w/2.0;
     this->dppar.tolerance = 1e-6;
     this->dppar.nx = int(this->dppar.Lx/h + 0.5);
```
Comment (rykerfish) [inline solvers/DPStokes/mobility.h:59]: Actually, I talked with Brennan and we think this line might be incorrect since it effectively changes the value for $h$ which affects the hydrodynamic radius. Since this is the code I used to regenerate the data, I'm going to look into this before finalizing the DPStokes parameters.

**Comment by RaulPPelaez** [comment]
I am not sure I am following. In these plots, you are comparing the GPU impl from the article with the GPU impl in libmobility?.
What this tells us is that the GPU code has not changed behavior since the publication of the article. Which is nice, but it might still have an issue, right? Or you say you do not trust the CPU implementation?
If you are comparing GPU with GPU then any error you see can probably be attributed to non deterministic execution order.

BTW one difference between CPU and GPU impls is that in the CPU particlesare spread to the cell corners in XY (pos_cell(i) = h*i), but in GPU I spread to cell centers (pos_cell(i) = (i+0.5)*h). This will give you differences in the results due to the variation of the hydrodynamic radius inside a cell. You should be able to accommodate for this by comparing CPU(particle_x, particle_y) with GPU(particle_x-0.5h, particle_y-0.5h).

**Comment by RaulPPelaez** [comment]
Also you should run all these correctness tests in double, numerical accuracy could be hiding some effects. For instance, the correction expressions are really bad for float as is and are written in a really specific way to reduce numerical issues. I bet the increased error you see near the wall is due to this.

**Comment by rykerfish** [comment]
@RaulPPelaez sorry about the delay in getting this finished- I was traveling for a bit, got busy with another project, and Brennan and I had spent some extra time talking about the DPStokes parameters til we felt comfortable with it.

What we landed on was leaving $h$ and $\beta$ at the values determined from the DPStokes paper as Brennan wanted to best preserve the hydrodynamic radius due to the high level of accuracy with which a particle's radius can be measured in experiments. A fixed value for $h$ requires altering the periodic length $L$ to make $N=L/h$ an integer for the FFT grid, which is now what is done in the libMobility parameter selection. 

One consequence of this is that the reference results from the DPStokes repository could no longer be used exactly since the hydrodynamics are a bit sensitive to $L$, so I regenerated reference results from this repository with the new parameters.

If this method of selection seems good, I think this branch is _finally_ good to go.

File: tests/test_mobility_matrix.py [hunk 3]
```diff
@@ -35,8 +35,8 @@ def test_mobility_matrix(Solver, periodicity, hydrodynamicRadius, numberParticle
     assert M.dtype == precision
     sym = M - M.T
     assert np.allclose(
-        sym, 0.0, rtol=0, atol=1e-7
-    ), f"Mobility matrix is not symmetric within 1e-6, max diff: {np.max(np.abs(sym))}"
+        sym, 0.0, rtol=0, atol=5e-5
```
Comment (rykerfish) [inline tests/test_mobility_matrix.py:38]: I want to specifically point out that the changes in the DPStokes parameter selection made the mobility matrix less symmetric. The DPStokes tests now only pass with a tolerance of 5e-5, as compared to the previous 1e-7.

Comment (brennansprinkle) [reply tests/test_mobility_matrix.py:38]: Hmmmm that's worth figuring out. Can you isolate *which* changes to the parameter selection made it less symmetric? Somehow I suspect it's the z-grid but it would also make sense it is was the kernel parameters

**Comment by RaulPPelaez** [comment]
All the contrary man, thank you so much for working on this.

Have you generated the reference using libMobility now then? That is dangerous bc it assumes what you are trying to test is correct beforehand.

**Comment by RaulPPelaez** [comment]
What we settled in the past with Aleks on this "you have to change r_h or L" is to have a parameter that allows you to select if L can be changed or not.

Note that by changing L you also change r_h due to periodic corrections.

**Comment by rykerfish** [comment]
I did regenerate the DPStokes references using libMobility, although I was able to use the old NBody references. 

What I did to have confidence when regenerating the reference was to first match the parameters used to generate the original reference and see that I matched the original reference to around $10^{-6}$ or so. Then, I'd change the parameters back to the ones I implemented and regenerate. Obviously not 100% bulletproof, though.

If we want to have both a parameter selection that preserves $L$ and one that is allowed to modify $L$, I could finish the implementation in #19 that emulates the fancy parameter selection. That should allow us to use the old reference data too by using the same parameter selection code that they used to generate the original refs. Do you think that's a good route to go?

Good point about modifying $L$ still changing $r_h$. Maybe our hope was that the effects from the periodic corrections could be minimized if necessary by making $L$ larger? Or maybe that's just the sign that we'll have to settle for changing something that effects $r_h$ in the end, like you said.

File: tests/test_mobility_matrix.py (hunk 3 above)
Comment (rykerfish) [reply tests/test_mobility_matrix.py:38]: I think it's both the z-grid and $m$. In this branch, I switched from $m=6$ to $m=4$ as we don't have torques implemented yet and that's what the paper recommends. Switching from 4 to 6 improves the worst-case error in how symmetric the mobility matrix is by 1-2 orders of magnitude, and doubling the number of points in the z-grid improves by another 1-2 orders of magnitude.

**Comment by RaulPPelaez** [comment]
What you did makes sense, I was not getting the full picture.

My advise here is to keep it simple for now. Add a parameter called "allowChangingBoxSize" or smth like that. If set to true (you can make this the default if you decide) change L so you can fix h, otherwise keep h fixed to the nearest possible value (changing r_h).

If the latter is too troublesome then just make the module change L and document accordingly. We can add that functionality when needed in the future.

**Comment by rykerfish** [comment]
@RaulPPelaez I've followed your suggestion and implemented a mode that changes $h$ while keeping $L$ fixed (as the new default), but I did it with the inverse-polynomial interpolation used in the DPStokes repository to re-select $\beta$ according to the new $h$. Theoretically, this should better preserve the relationship between $r_h$, $h$, and $\beta$. 

I added the parameter to select which mode to use- I did it in setParameters() since I thought it made sense to go with where you set $L$, but it could get moved to initialize() if you think that's a better place. 

Finally, I regenerated the ref data for hopefully the last time with the new default. Once again, I compared against the data from the original DPStokes repository to verify a decent match before regenerating.

**Comment by RaulPPelaez** [comment]
Sounds good Ryker. Is this ready to merge?

**Comment by rykerfish** [comment]
@RaulPPelaez Ready to merge unless you find anything to change on review!

File: solvers/DPStokes/extra/poly_fits.h [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    double accumulator = polyCoeffs[order];
+    double current_x = x;
+    for(int i = 1; i <= order; i++){
+      accumulator += polyCoeffs[order-i]*current_x;
+      current_x *= x;
+    }
+
+    return accumulator;
+  }
+
```
Comment (RaulPPelaez) [inline solvers/DPStokes/extra/poly_fits.h:17]: It would be great to have some comment here about what this is and where it comes from. Maybe a header in this file.

File: solvers/DPStokes/mobility.h [hunk 5]
```diff
@@ -51,23 +52,58 @@ class DPStokes: public libmobility::Mobility{
     this->lanczosTolerance = ipar.tolerance;
     this->dppar.mode = this->wallmode;
     this->dppar.hydrodynamicRadius = ipar.hydrodynamicRadius[0];
-    this->dppar.w = 6;
-    this->dppar.beta = 1.714*this->dppar.w;
-    real h = this->dppar.hydrodynamicRadius/1.554;
+    this->dppar.w = 4;
```
Comment (RaulPPelaez) [inline solvers/DPStokes/mobility.h:55]: Same here, add a little comment with context

File: solvers/DPStokes/mobility.h [hunk 6]
```diff
@@ -51,23 +52,58 @@ class DPStokes: public libmobility::Mobility{
...
+    int N;
+    // either N_up or N_down will be a multiple of 2. pick the even one for a more FFT friendly grid.
+    if(N_up % 2 == 0){
+      N = N_up;
+    }else{
+      N = N_down;
+    }
+
+    this->dppar.nx = N;
+    this->dppar.ny = N;
```
Comment (RaulPPelaez) [inline solvers/DPStokes/mobility.h:73]: You can rewrite as:
```c++
    int N = floor(this->dppar.Lx / h);
    N+= N % 2;
    this->dppar.nx = this->dppar.ny = N;
```
BTW, is the fact that Ly is ignored on purpose? Are we supporting only Lx=Ly or something like that?

File: solvers/DPStokes/mobility.h [hunk 7]
```diff
@@ -51,23 +52,58 @@ class DPStokes: public libmobility::Mobility{
...
+
+    // note: only set up for square boxes
```
Comment (RaulPPelaez) [inline solvers/DPStokes/mobility.h:75]: If this is a real restriction make it throw an exception when the use sends Lx!=Ly
```c++
if(this->dppar.Lx != this->dppar.Ly)
    throw std::runtime_error("[DPStokes] Only square boxes (Lx==Ly) are currently supported");
```

File: solvers/DPStokes/mobility.h [hunk 8]
```diff
@@ -51,23 +52,58 @@ class DPStokes: public libmobility::Mobility{
...
+
+    // pick nearby N such that 2(Nz-1) is FFT friendly
+    N_up = ceil(nz_actual);
+    N_down = floor(nz_actual);
+    if(N_up % 2 == 1){
+      this->dppar.nz = N_up;
+    } else {
+      this->dppar.nz = N_down;
+    }
+
```
Comment (RaulPPelaez) [inline solvers/DPStokes/mobility.h:106]: Same as above

File: tests/test_fluctuation_dissipation.py [hunk 9]
```diff
@@ -62,7 +62,7 @@ def fluctuation_dissipation_KS(M, fluctuation_method):
     ],
 )
 @pytest.mark.parametrize("hydrodynamicRadius", [0.95, 1.12])
-@pytest.mark.parametrize("numberParticles", [1, 2, 3, 10])
+@pytest.mark.parametrize("numberParticles", [10])
```
Comment (RaulPPelaez) [inline tests/test_fluctuation_dissipation.py:65]: Introduce 1 and 2. Bugs arise at 1 because it tends to be a corner case.
If it fails at 2 it is easier to debug than at 10.

File: tests/test_mobility_matrix.py [hunk 10]
```diff
@@ -35,8 +35,8 @@ def test_mobility_matrix(Solver, periodicity, hydrodynamicRadius, numberParticle
...
+    ), f"Mobility matrix is not symmetric within 5e-5, max diff: {np.max(np.abs(sym))}"
```
Comment (RaulPPelaez) [inline tests/test_mobility_matrix.py:39]: You should not hardcode this 5e-5. Just use atol:
```python
f"Mobility matrix is not symmetric within {atol}, max diff: {np.max(np.abs(sym))}"
```

**Review by RaulPPelaez — COMMENTED** [review]
I have some nitpicks and comments, but its great. Thanks Ryker.

File: solvers/NBody/extra/hydrodynamicKernels.cuh [hunk 11]
```diff
@@ -70,9 +70,9 @@ namespace nbody_rpy{
 	real invZi = real(1.0) / hj;
 	real invZi3 = invZi * invZi * invZi;
 	real invZi5 = invZi3 * invZi * invZi;
-	correction.x += -vj.x*(real(9.0)*invZi - real(2.0)*invZi3 + invZi5 ) / real(12.0);
-	correction.y += -vj.y*(real(9.0)*invZi - real(2.0)*invZi3 + invZi5 ) / real(12.0);
-	correction.z += -vj.z*(real(9.0)*invZi - real(4.0)*invZi3 + invZi5 ) / real(6.0);
```
[the remark this answers is not in the document]
Comment (brennansprinkle) [reply solvers/NBody/extra/hydrodynamicKernels.cuh:75]: Haha! Agreed, this was a good catch!

File: tests/test_wall_mobility.py [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+        (NBody, ("open", "open", "single_wall"), "self_mobility_bw_ref_noimg.mat")
+    ],
+)
+def test_self_mobility(Solver, periodicity, ref_file):
+    zmax = 19.2
+    xymax = 76.8
+    params = self_mobility_params[Solver.__name__]
+
+    ref_dir = "./ref/"
+    ref = scipy.io.loadmat(ref_dir + ref_file)
```
[the remark this answers is not in the document]
Comment (brennansprinkle) [reply tests/test_wall_mobility.py:29]: Also agree that we should remove scipy as a dependency whenever possible - it's a buggy, bad library

File: solvers/DPStokes/extra/poly_fits.h (hunk 4 above)
Comment (brennansprinkle) [reply solvers/DPStokes/extra/poly_fits.h:17]: Yeah I agree with this as well.

Comment (brennansprinkle) [reply solvers/DPStokes/extra/poly_fits.h:17]: Also for some reason it won't let me reply to your comments about the symmetry check - but that's a good point about the relative precision in the symmetry. Though like you said, it's probably best to do a mix of relative error (when the value is above e.g 1e-4) and absolute error (when the value is below that threshold)

File: tests/test_mobility_matrix.py (hunk 3 above)
Comment (rykerfish) [reply tests/test_mobility_matrix.py:38]: Testing in double & single creates no noticeable difference, so I think you're right that it's just a tolerance issue.

File: solvers/DPStokes/mobility.h (hunk 6 above)
Comment (rykerfish) [reply solvers/DPStokes/mobility.h:73]: Good catch- I originally wrote this supporting only Lx=Ly as the parameter selection code from DPStokes did that. I think @brennansprinkle mentioned it was likely because having Nx=Ny is the best for efficiency reasons related to FFTs, but I might be remembering incorrectly. 

This routine could be modified to support Lx != Ly. We could try this if it isn't too complicated:

1. For a general Lx and Ly, try and compute as before leaving the box size unchanged. This will produce an $h_x$ and $h_y$ that aren't going to be the same, in general. 
2. If $h_x = h_y$, proceed as normal. If not, throw an error that says you must either set Lx = Ly or allow us to change the box size. If Lx = Ly, then we're guaranteed that $h_x = h_y$. If we're allowed to change the box size, we let $h = \min(h_x, h_y)$ and then modify one of Lx or Ly to make Nx= Lx/h and Ny = Ly/h integers. This would give us a uniform grid spacing but with a different number of nodes in each direction.

Should I implement the above parameter selection or simply require that Lx = Ly?

File: tests/test_wall_mobility.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (rykerfish) [reply tests/test_wall_mobility.py]: I converted everything in the wall tests to use numpy, but it's worth mentioning that the fluctuation dissipation test depends on scipy.

[the remark this answers is not in the document]
Comment (brennansprinkle) [reply tests/test_wall_mobility.py]: lol good point, but it's still good to *reduce* our dependence on garbage :)

**Comment by rykerfish** [comment]
@RaulPPelaez This PR should be ready to merge- I think I addressed all your comments but let me know if there is anything else you notice!

I talked with Brennan and we're deferring implementing rectangular boxes (Lx != Ly) for DPStokes to a later date, so I put an error in there during initialization.

File: tests/test_fluctuation_dissipation.py [hunk 13]
```diff
@@ -62,7 +62,8 @@ def fluctuation_dissipation_KS(M, fluctuation_method):
     ],
 )
 @pytest.mark.parametrize("hydrodynamicRadius", [0.95, 1.12])
-@pytest.mark.parametrize("numberParticles", [1,2,10])
+# @pytest.mark.parametrize("numberParticles", [1,2,10])
+@pytest.mark.parametrize("numberParticles", [10])
```
Comment (RaulPPelaez) [inline tests/test_fluctuation_dissipation.py:66]: Why did you removed the 1,2 again? Is it failing or just takes a lot of time?

Comment (rykerfish) [reply tests/test_fluctuation_dissipation.py:66]: Whoops, forgot to revert that despite looking through the files! That test is quite slow for DPStokes so I often run only the 10 particle case if I'm running through tests more than once.

**Comment by RaulPPelaez** [comment]
> I talked with Brennan and we're deferring implementing rectangular boxes (Lx != Ly) for DPStokes to a later date, so I put an error in there during initialization.

That's fine, we will get around it.

File: tests/test_fluctuation_dissipation.py (hunk 13 above)
Comment (RaulPPelaez) [reply tests/test_fluctuation_dissipation.py:66]: One can mark tests or individual parameters with custom markers:
```python
@pytest.mark.parametrize(
    "numberParticles", [1, 2, pytest.param(10, marks=pytest.mark.slow)]
)
```
Then you can run with `pytest -m slow tests*py` or `pytest -n "not slow" tests*py`.
Technically you have to register these in a separate file, but using them as is just works for me https://docs.pytest.org/en/stable/how-to/mark.html

Something to keep in mind for the future, not worth it at this point IMO.

**Review by RaulPPelaez — APPROVED** [review]
Thanks for your hard work Ryker!
